pkg/controller/endpoint/endpoints_controller_cache.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoint
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+// EndpointsControllerCache is a util that manages a local copy of the state of pods and services
+// for purpose of endpoints object syncing.
+//
+// The purpose of this util is to create a single source of truth that is thread-safe and avoids
+// potential race-conditions in the endpoints-controller. In addition it allows to easily determine
+// which object (pod/service) changes resulted in the endpoints object change, which is crucial in
+// implementing the network programming latency metric.
+//
+// The cache should be initialized via pod and service listers and kept up-to-date via the
+// (Add|Update|Delete)(Pod|Service) methods called in the EndpointsController's event handlers.
+// The Initialize method together with the event handlers method are guarded by the same mutex which
+// guarantees that no information is ever missed and all events are processed in the serial order.
```

Comment:
I think this is eventually correct, but one more thing we probably want to add comment is that:
- given that you register event handlers early in the endpoints controller, you don't have full control whether:
(a) first callback from informer or
(b) cache.initialize() will be run (this one is called from Run() method).

So it may happen, that you already received a set of callbacks from the informer, the initialize is called, and then the rest of callbacks is being delivered. However, given that the state in lister is updated before any callback is send, that should be fine.
But we should add a comment about that.

[The more I think about it though, the more I would like to the initialization to be based on callbacks - let me try to hack something around a little bit]

Reply:
Yes, the reason it works is that the "Initialize" and Informer events are mutually exclusive and list always sees everything that has been already delivered via events.

Updated the doc of the Initialize method, but after some thinking I removed the section saying that this initialization mechanism depends on getting initial OnAdd events for all objects. It's actually not necessary in this implementation. As long as these two assumptions are fulfilled
1. Informer events and Initialize() are mutually exclusive 
2. Initialize() sees everything (at least, can see more) that has been already delivered via informer events

then we are guaranteed that the cache won't miss anything. 

That said, OnAdds are still required in the endpoints_controlller implementation, as it guarantees that when endpoints_controller starts all services will be added to the queue and initially resynced.

PTAL